packages/fx-core/src/core/middleware/projectUpgrader.ts

```diff
@@ -94,41 +117,47 @@ export async function upgradeContext(ctx: CoreHookContext): Promise<void> {
     // Some keys updated. Send start telemetry.
     sendTelemetryEvent(core?.tools?.telemetryReporter, inputs, TelemetryEvent.ProjectUpgradeStart);
 
-    // Read UserData file.
-    const userDataPath = path.resolve(confFolderPath, `${defaultEnvName}.userdata`);
-    const userData = await readUserData(userDataPath, projectSettings.projectId);
-
     // Merge updatedKeys into UserData.
     mergeKeysToUserDate(userData, updatedKeys);
+  } catch (error) {
+    const errorObject = ContextUpgradeError(error, error.message == SolutionContextNotFoundError);
+    core?.tools?.logProvider?.info(
+      `Template upgrade failed. Please clean the env.default.json and default.userdata file and try again. Reason: ${error?.message}`
+    );
+    sendTelemetryErrorEvent(
+      core?.tools?.telemetryReporter,
+      inputs,
+      TelemetryEvent.ProjectUpgrade,
+      errorObject
+    );
+    ctx.result = err(errorObject);
+    return;
+  }
 
+  try {
     // Save the updated context and UserData.
     await saveContext(contextPath, context);
     await saveUserData(userDataPath, userData, projectSettings.projectId);
-
-    // Send log.
-    core?.tools?.logProvider?.info(
-      "[core]: template version is too low. Updated context and moved some configs from env to userdata."
-    );
-    sendTelemetryEvent(core?.tools?.telemetryReporter, inputs, TelemetryEvent.ProjectUpgrade, {
-      [TelemetryProperty.Success]: TelemetrySuccess.Yes,
-    });
   } catch (error) {
-    const errorObject = ContextUpgradeError(error);
-    core?.tools?.logProvider?.info(
-      `Template upgrade failed. Please clean the env.default.json and default.userdata file and try again. Reason: ${error?.message}`
-    );
+    const errorObject = WriteFileError(error);
+    core?.tools?.logProvider?.info(errorObject.message);
     sendTelemetryErrorEvent(
       core?.tools?.telemetryReporter,
       inputs,
       TelemetryEvent.ProjectUpgrade,
-      errorObject,
-      {
-        [TelemetryProperty.Success]: TelemetrySuccess.No,
-        [TelemetryProperty.ErrorMessage]: error?.message,
-      }
+      errorObject
```

Comment:
why this part remove:
```
  {
        [TelemetryProperty.Success]: TelemetrySuccess.No,
        [TelemetryProperty.ErrorMessage]: error?.message,
      }
```

Reply:
It is included in sendTelemetryErrorEvent:
![image](https://user-images.githubusercontent.com/63089166/128966808-6088d331-f0f5-41d5-bf65-a29b23888ce1.png)